parser/wasm_parse/mod.rs

```diff
@@ -15,50 +15,125 @@ pub struct SectionIndices {
     globals: Vec<Id>,
 }
 
-struct IndexedSection<'a>(usize, wasmparser::Section<'a>);
+struct IndexedSection<'a>(usize, wasmparser::Payload<'a>);
 
-impl<'a> Parse<'a> for wasmparser::ModuleReader<'a> {
+struct CodeSection<'a> {
+    index: usize,
+    reader: wasmparser::CodeSectionReader<'a>,
+    byte_size: usize,
+}
+
+struct FunctionSection<'a> {
+    index: usize,
+    reader: wasmparser::FunctionSectionReader<'a>,
+    byte_size: usize,
+}
+
+pub struct ModuleReader<'a> {
+    data: &'a [u8],
+    offset: usize,
+    parser: wasmparser::Parser,
+}
+
+impl<'a> ModuleReader<'a> {
+    pub fn new(data: &[u8]) -> ModuleReader {
+        ModuleReader {
+            data: data,
+            offset: 0,
+            parser: wasmparser::Parser::new(0),
+        }
+    }
+
+    fn current_position(&self) -> usize {
+        self.offset
+    }
+
+    fn eof(&self) -> bool {
+        self.offset == self.data.len()
+    }
+
+    fn read(&mut self) -> Result<wasmparser::Payload<'a>, traits::Error> {
+        let (section, bytes_consumed) =
+            match self.parser.parse(&self.data[self.offset..], self.eof())? {
+                wasmparser::Chunk::NeedMoreData { .. } => {
+                    panic!("@@@ nyi");
```

Comment:
Let's return an error here rather than panic since the function is already set up for that